include/test/mir/test/event_matchers.h

```diff
@@ -32,6 +34,195 @@
+-> char const*;
+auto get_enum_value(MirTouchAxis axis)
+-> char const*;
+auto get_enum_value(MirTouchTooltype tooltype)
+-> char const*;
+auto get_enum_value(MirPointerAction action) 
+-> char const*;
+auto get_enum_value(MirPointerAxis axis)
+-> char const*;
+auto get_enum_value(MirPointerButtons button)
+-> char const*;
+auto get_enum_value(MirPointerAxisSource source)
+-> char const*;
+
+/// Takes in two enums with get_enum_value() implemented and determines if they are the same.
+/// If false, the discrepancy is logged to the MatchResultListener.
+template<typename T> auto inline enums_match(
+    T const& expected,
+    T const& actual,
+    testing::MatchResultListener* result_listener)
+-> bool
+{
+    auto type_name = boost::typeindex::type_id<T>().pretty_name();
+    if (expected != actual)
+    {
+        *result_listener << "Expected " << type_name << " (" << get_enum_value(expected) << ") does not match actual (" << get_enum_value(actual) << ")";
+        return false;
+    }
+
+    return true;
+}
+
+/// If false, the discrepancy is logged to the MatchResultListener.
+auto inline touch_ids_match(
+    MirTouchId const& expected,
+    MirTouchId const& actual,
+    testing::MatchResultListener* result_listener)
+-> bool
+{
+    if (expected != actual)
+    {
+        *result_listener << "Expected MirTouchId (" << expected << ") does not match actual (" << actual << ")";
+        return false;
+    }
+
+    return true;
+}
+
+/// Takes in a Mir*EventModifiers and an int bitmask and determines if they are the same.
+/// If not, the discrepancy is logged to the MatchResultListener.
+template<typename T> auto inline modifiers_match(
+    T const& expected,
+    int actual,
+    testing::MatchResultListener* result_listener)
+-> bool
+{
+    if (expected != actual)
+    {
+        *result_listener << "Expected modifiers (" << expected << ") do not match actual (" << actual << ")";
+        return false;
+    }
+
+    return true;
+}
+
+// If false, the discrepancy is logged to the MatchResultListener.
+auto inline keysyms_match(
+    xkb_keysym_t expected,
+    xkb_keysym_t actual,
+    testing::MatchResultListener* result_listener)
+-> bool
+{
+    if (expected != actual)
+    {
+        *result_listener << "Expected keysym (" << expected << ") does not match actual (" << actual << ")";
+        return false;
+    }
+
+    return true;
+}
+
+// If false, the discrepancy is logged to the MatchResultListener.
+auto inline scan_codes_match(
+        int expected,
+        int actual,
+        testing::MatchResultListener* result_listener)
+-> bool
+{
+    if (expected != actual)
+    {
+        *result_listener << "Expected scan code (" << expected << ") does not match actual (" << actual << ")";
+        return false;
+    }
+
+    return true;
+}
+
+// If false, the discrepancy is logged to the MatchResultListener.
+auto inline button_states_match(
+    bool expected,
+    bool actual,
+    testing::MatchResultListener* result_listener)
```

Comment:
Ah, this was because I wasn't aware that both `MirPointerButton` and `MirPointerButtons` existed.

Reply:
By creating buttons_states_match(), I think this removes the need for the overload prevention. Let me know if this is incorrect.

Reply:
> By creating buttons_states_match(), I think this removes the need for the overload prevention. Let me know if this is incorrect.

The "overload protection" is needed because things implicitly convert to `bool` too easily. That makes undiagnosed mistakes possible (as you inadvertently demonstrated).

Having a different function with a similar name doesn't address that problem